design/one-pager-template-stacks.md

```diff
@@ -0,0 +1,679 @@
+needed permissions.
+
+#### Concerning the use of Core Types
+
+In several examples throughout this design, Template Stacks are demonstrated
+generating `Deployment` resources by including them in template text. There are
+compelling justifications for both permitting and denying Template Stacks from
+declaring a dependency on core and extension types, such as
+`deployment.apps/v1`.
+
+The ability to use Template Stacks to render arbitrary Kubernetes resources is a
+powerful feature. Such a privilege would change the boundaries of a Stack,
+permitting it to function outside of Crossplane managed types. Template Stacks
+would compete with `KubernetesApplication` on this capability.
+
+If Template Stacks are prohibited from using core and extension types, the
+complement of `KubernetesApplication` and a local `KubernetesCluster` can be
+used to overcome the limitation while providing additional security boundaries.
+
+When considering that the Stack-Manager does not depend on Crossplane, nor
+Crossplane on the Stack-Manager (strictly speaking), we can see that Template
+Stacks, and the Stack Manager, may have general utility outside of Crossplane.
+When used with Crossplane, the Stack Manager administrator may desire a more
+opinionated approach, restricting Template Stacks to use on known types,
+specifically Crossplane managed types.
+
+These are features and challenges to investigate in a future design.
+
+### Life Cycle and Updates
+
+#### Resource life cycle
+
+On each reconciliation loop for a TSM managed kind, the rendering process of the
+TSM will use a single pass at applying templates.
+
+* Get the current resource
+* Get the `templates` from the `Stack` resource that owns this CRD
+* All of the dependent resources (generated from templates) are fetched.
+* The current resource and dependent resources are combined to form variables to
+  pass to the template.
+* Template substitutions are made for each resource.
+* All of the dependent, templated, resources are applied to the API, including a
+  pass at deleting resources whose templates result is empty.
+
+Because this process operates in a single read then write pass, it can not get
+caught in a circular dependency loop. However, Template Stack creators should
+avoid crafting their templates in a way that they would never resolve as
+intended.
+
+The rendered template will replace changes to dependent resources that are found
+in the Kubernetes API. More importantly, since the values are obtained directly
+from the API on each pass, if a resource is updated by a user or some other
+process, those values will be used when rendering.
+
+If a templates rendering contains invalid resource YAML, or contains fields and
+values are not valid for the resource being generated, the Template Stack
+Manager will need a defined means of surfacing these error conditions. Just as
+sibling resources should be made available through templates, the error messages
```

Comment:
I'm not sure I'm following what this means to make error messages available through templates.  Is an example helpful here, or maybe another sentence or two describing this functionality?

Is it in scope for this intial effort?

Reply:
The set of variables (and/or functions) to present to a template should include:
* the Stack managed object (the `Foo`)
* the set of dependent resource objects (the `DependentThing`)
* the set of errors encountered rendering each dependent resource
  This will allow `Foo` to reflect errors rendering `DependentThing` as a conditioned status using `templateStatus`.
* helper functions for string manipulation, list handling, and simple numeric operations, such as https://github.com/leekchan/gtf#index or https://github.com/hashicorp/consul-template/

Additionally, we may want to expose
* the errors encountered rendering the templateStatus
* the raw templates themselves.
   This may be useful for storing partial templates in the Stack.  The key name would not match a GVK in this case, which means it could be an arbitrary identifier.  The key could resemble a filepath, which could help in the migration of Helm chart "includes" if an "include" function is provided in the template engine.
   ```yaml
   kind: Stack
   spec: { name: "Foo" }
   templates:
     this/name/can/be/anything:
       so/can/this/name: |
         Reusable partial template {{.someVar}}.
         The keys do not have to match GVKs owned by this Stack,
         permitting this use case.
     foo.group/version:
       resourceTemplate:
          kind: DependentThing
          spec:
             text: {{ include "this/name/can/be/anything" "so/can/this/name" }}
   ```

Reply:
This text has been included in the design.